ansible-runner integration

chrismeyersfsu opened the pull request:
Instead of using pexpect directly, use ansible-runner to launch playbooks.

chrismeyersfsu committed:
init ansible-runner requirements

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -1374,14 +1374,13 @@ def awx_meta_vars(self):
+                r['{}_workflow_job_id'.format(name)] = wj.pk
+                r['{}_workflow_job_name'.format(name)] = wj.name
 
+        if not created_by:
```
looks like these changes are just due to a rebase problem??

chrismeyersfsu replied:
?

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -769,43 +770,50 @@ def build_passwords(self, instance, **kwargs):
         return env
 
-    def build_env(self, instance, **kwargs):
+    def build_env(self, instance, private_data_dir, isolated, private_data_files={}):
```
mutable default arguments are frowned upon, but shouldn't matter here since it's only read, never changed.

ryanpetrello replied:
It's only read _at this moment_.  Let's default this to `None` and do the defaulting in the method so we don't introduce some subtle bug later.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -822,41 +830,57 @@ def build_env(self, instance, **kwargs):
         return False
 
-    def build_inventory(self, instance, **kwargs):
+    def build_inventory(self, instance, private_data_dir):
```
I've been going through these, and all of these line up with what I was expecting, and using within the inventory source injector stuff. So that's good.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -769,43 +770,50 @@ def build_passwords(self, instance, **kwargs):
+            if e.errno != errno.EEXIST:
+                raise
+
+        path = os.path.join(env_path, 'extravars')
```
so we need it to be specifically `env/extravars`, as per the ansible-runner examples, that makes sense.

https://ansible-runner.readthedocs.io/en/latest/intro.html#env-extravars
https://github.com/ansible/ansible-runner/blob/master/demo/env/extravars

I've been mentioning a lot to @ryanpetrello that it seems like we're going to be going further away from the randomized file names, and here is a case where ansible-runner looks like it necessitates doing so... similar to how the plugins require a certain filename for inventory.

One thing I do not get, however, didn't we have some cases where we passed _multiple_ `-e @file1.yml -e @file2.yml` kind of arguments to `ansible-playbook`? Given the ansible-runner demo, I do not see how it allows that.

I can swear I remember us doing this. Ansible, itself, will combine the variables given in multiple `-e` flags, and that's very useful functionality.

chrismeyersfsu committed:
use ansible runner to run playbooks

* Project Updates
* Jobs
* Inventory Updates
* System Jobs
* AdHoc Commands

* Notifications
* Fact Cache
* proot

ryanpetrello committed:
update isolated task execution for ansible-runner

chrismeyersfsu committed:
remove safe_args and add status_handler

* safe_args no longer makes sense. We have moved extra_vars to a file
and thus do not pass sensitive content on the cmdline

shanemcd commented on Makefile:
```diff
@@ -146,6 +146,7 @@ virtualenv_awx:
 		fi; \
 		if [ ! -d "$(VENV_BASE)/awx" ]; then \
 			$(PYTHON) -m venv $(VENV_BASE)/awx; \
+			$(VENV_BASE)/awx/bin/pip install $(PIP_OPTIONS) --ignore-installed docutils==0.14; \
```
Can we move this into a requirements file?

chrismeyersfsu committed:
refactor and fix unit tests

* fixup task TestGenericRun
* make runner callback functions accessable to testing
* reduce isintance() usage in run() by using build_ pattern
* move process_isolation param building to build_ function so it can be
tested

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -797,23 +801,36 @@ def build_passwords(self, instance, **kwargs):
+    def build_extra_vars_file(self, instance, private_data_dir, passwords):
+        '''
+        Build ansible yaml file filled with extra vars to be passed via -e@file.yml
+        '''
```
Is this method intentionally empty?

chrismeyersfsu replied:
yeah

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -952,184 +954,247 @@ def run(self, pk, **kwargs):
+                proot_custom_virtualenv = self.instance.ansible_virtualenv_path
+            self.build_extra_vars_file(self.instance, private_data_dir, passwords)
+            args = self.build_args(self.instance, private_data_dir, passwords)
+            # TODO: output_replacements hurts my head right now
```
Do we still need this to sanitize scm passwords in project update stdout?  I've always been a little fuzzy on why this code existed here in the first place.

ryanpetrello replied:
Would like to see this answered/addressed pre-merge or shortly thereafter.

chrismeyersfsu replied:
Good question. I wanted to ask if anyone remembered the history of this. I recall it having to do with the git version requiring us to put the password in the url i.e. git://user:pass@url.com:gitrepo Maybe we don't support that version of git that comes with some distro or maybe some combo of Ansible version made us do this. So maybe we can remove it. Need to dig in.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1303,9 +1364,9 @@ def build_args(self, job, **kwargs):
-            ssh_username = kwargs.get('username', creds.get_input('username', default=''))
-            become_method = kwargs.get('become_method', creds.get_input('become_method', default=''))
-            become_username = kwargs.get('become_username', creds.get_input('become_username', default=''))
+            ssh_username = creds.username
```
Let's continue using `.get_input()`, this magic attribute access is going away in the near future.

AlanCoding replied:
yeah this is pretty important to resolve in the diff before merge

ryanpetrello replied:
Would like to have this addressed pre-merge.

ryanpetrello commented on requirements/requirements_git.txt:
```diff
@@ -1,3 +1,4 @@
 git+https://github.com/ansible/ansiconv.git@tower_1.0.0#egg=ansiconv
 git+https://github.com/ansible/django-qsstats-magic.git@py3#egg=django-qsstats-magic
 git+https://github.com/ansible/django-jsonbfield@fix-sqlite_serialization#egg=jsonbfield
+git+https://github.com/ansible/ansible-runner@master#egg=ansible_runner
```
Should we cut a new ansible-runner before merging this?

ryanpetrello commented on tools/docker-isolated/Dockerfile:
```diff
@@ -3,23 +3,22 @@ RUN yum clean all
 RUN ln -s /usr/bin/python36 /usr/bin/python3
 RUN python36 -m ensurepip
 RUN pip3 install virtualenv
+RUN pip3 install git+https://github.com/ansible/ansible-runner@master#egg=ansible_runner
```
Same question as above re runner release

ryanpetrello commented on awx/main/tests/unit/models/test_survey_models.py:
```diff
@@ -132,29 +132,6 @@ def test_survey_passwords_not_in_extra_vars():
     }
 
 
-def test_job_safe_args_redacted_passwords(job):
```
Are we still verifying this behavior elsewhere?

chrismeyersfsu replied:
It's not a thing. We don't pass extra vars via -e, we embed them in files and then -e the file.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -952,184 +954,247 @@ def run(self, pk, **kwargs):
-                self.update_model(pk, custom_virtualenv=getattr(instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV_PATH))
-
-            # Fetch ansible version once here to support version-dependent features.
-            kwargs['ansible_version'] = get_ansible_version()
```
Is there any way we can leave the calculation of ansible_version here? because I would just be adding it back.

This seems like a reasonable thing that could be passed around, maybe `self.ansible_version` would be better?

chrismeyersfsu committed:
point at another runner branch

* revert parent_uuid because it causes problems with unexpected
parameter on event creation for some event type.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -771,23 +771,27 @@ def build_private_data_files(self, instance, **kwargs):
-                    name = 'credential_%d' % credential.pk
-                    path = os.path.join(kwargs['private_data_dir'], name)
-                    run.open_fifo_write(path, data)
+                    try:
```
What's this about?  Because you're looping?

chrismeyersfsu replied:
bit of a safety net. Making the private_data_dir needs to happen early because everything gets dumped in there. Often times a feature is added that needs to do things earlier than other existing things. People often duplicate the creating of the private_data_dir. This protects against the private_data_dir already existing in that common scenario.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -901,235 +953,253 @@ def get_instance_timeout(self, instance):
         '''
-        instance = self.update_model(pk, status='running',
-                                     start_args='')  # blank field to remove encrypted passwords
+        # self.instance because of the update_model pattern and when it's used in callback handlers
```
👍

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -1360,14 +1422,25 @@ def build_args(self, job, **kwargs):
+        return cwd
+
+    def build_playbook_path_relative_to_cwd(self, job, private_data_dir):
+        return os.path.join(job.playbook)
```
is this needed at all? Seems like this is basically `lambda job: job.playbook`

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -901,235 +953,253 @@ def get_instance_timeout(self, instance):
+        Ansible runner callback triggered on status transition
+        '''
+        if status_data['status'] == 'starting':
+            self.instance = self.update_model(self.instance.pk, job_args=json.dumps(runner_config.command),
```
I know that safe_args wasn't needed due to the extra_vars anymore, are other forms of unsafe data handled by runner?

chrismeyersfsu replied:
Not that I'm aware of. Wanted to see if anyone else could think of any.

ryanpetrello reviewed (approved):
This looks good to me in spirit, some minor review comments/questions.

ryanpetrello commented:
After this is merged, @chrismeyersfsu we should run all of this through coverage and find the dead code.

chrismeyersfsu committed:
fixup unit tests for tasks

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -2071,16 +2100,20 @@ def build_passwords(self, inventory_update, **kwargs):
                 passwords[k] = credential.get_input(passkey, default='')
         return passwords
```
build passwords for inventory updates doesn't make any sense to me. Does anyone know why this existed?

chrismeyersfsu committed:
update tests

chrismeyersfsu commented:
recheck

chrismeyersfsu committed:
allow for runner setting parent_uuid

* Previously, parent_uuid was expected only on events generated for a
Job run. Now, there maybe a parent_uuid for any job type. AWX does not
support parenting events for any job type other than Job.

chrismeyersfsu committed:
do not save sensitive env vars

* job_env gets exposed via the api. Sensitive env variables should be
redacted before saved into job_env.

chrismeyersfsu committed:
fix more unit tests for runner

* isolated will be fixed in the future so pytest skip those
* fact cache moved one directory level up, account for that

chrismeyersfsu committed:
add license files for python modules

* python-daemon
* ansible-runner

chrismeyersfsu committed:
flake8 fix

chrismeyersfsu commented:
recheck

chrismeyersfsu commented:
recheck

chrismeyersfsu committed:
change from runner master to runner 1.3

chrismeyersfsu commented:
recheck

chrismeyersfsu commented:
recheck

chrismeyersfsu committed:
doc update fix
